File: python/compute_ACscore.py
import numpy as np

def compute_ACscore(labels, markers, positive_class):
    """
    Compute the ACscore metric for binary classifiers.

    Parameters:
    labels: Groundtruth labels (list, numpy array, or similar, 1D)
    markers: Predicted labels or scores (same format as labels)
    positive_class: Value representing the positive class

    Returns:
    ACscore: The ACscore evaluation result.
    ACscore_array: ACscore evaluations for possible cuts (if markers are scores).
    best_cut: Cut corresponding to the highest ACscore (if markers are scores).
    direction: 1 if scores >= best_cut are positive; 0 otherwise.
    best_predicted: Predicted labels for the highest ACscore (if markers are scores).
    """
    labels = np.asarray(labels)
    markers = np.asarray(markers)

    is_marker_score = not np.array_equal(np.unique(markers), [0, 1])

    if labels.ndim != 1 or markers.ndim != 1:
        raise ValueError("Input labels and markers must be 1D arrays.")

    unique_labels = np.unique(labels)
    if len(unique_labels) != 2:
        raise ValueError("The number of classes must be 2.")

    if positive_class not in unique_labels:
        raise ValueError("The specified positive class is incorrect.")

    negative_class = unique_labels[unique_labels != positive_class][0]

    mapped_labels = (labels == positive_class).astype(int)

    if not is_marker_score:
        mapped_markers = (markers == positive_class).astype(int)
        tp, tn, fp, fn = compute_confusion_matrix(mapped_labels, mapped_markers)
        ACscore = compute_ac_score(tp, tn, fp, fn)
        return ACscore, None, None, None, None

    # If markers are scores
    sorted_indices = np.argsort(markers)
    markers_sorted = markers[sorted_indices]
    labels_sorted = mapped_labels[sorted_indices]

    ACscore_array_forward, ACscore_forward, idx_max_forward = compute_ACscore_unidirectional(labels_sorted, "forward")
    ACscore_array_backward, ACscore_backward, idx_max_backward = compute_ACscore_unidirectional(labels_sorted, "backward")

    if ACscore_forward >= ACscore_backward:
        ACscore_array = ACscore_array_forward[np.argsort(sorted_indices)]
        ACscore = ACscore_forward
        idx_max = idx_max_forward
        direction = 1
    else:
        ACscore_array = ACscore_array_backward[np.argsort(sorted_indices)]
        ACscore = ACscore_backward
        idx_max = idx_max_backward
        direction = 0

    delta = 1e-6
    if idx_max == 0:
        best_cut = markers_sorted[0] - delta
    elif idx_max == len(markers) + 1:
        best_cut = markers_sorted[-1] + delta
    else:
        best_cut = (markers_sorted[idx_max - 1] + markers_sorted[idx_max]) / 2

    if direction == 1:
        best_predicted = np.concatenate((
            np.full(idx_max - 1, negative_class),
            np.full(len(markers) - idx_max + 1, positive_class)
        ))
    else:
        best_predicted = np.concatenate((
            np.full(idx_max - 1, positive_class),
            np.full(len(markers) - idx_max + 1, negative_class)
        ))

    return ACscore, ACscore_array, best_cut, direction, best_predicted

def compute_confusion_matrix(labels, predicted):
    tp = np.sum((predicted == 1) & (labels == 1))
    tn = np.sum((predicted == 0) & (labels == 0))
    fp = np.sum((predicted == 1) & (labels == 0))
    fn = np.sum((predicted == 0) & (labels == 1))
    return tp, tn, fp, fn

def compute_ac_score(tp, tn, fp, fn):
    total = tp + tn + fp + fn
    if tp + tn == total:
        return 1
    if fp + fn == total:
        return 0
    tp_tn = tp * tn
    return (2 * tp_tn) / (2 * tp_tn + tp * fp + tn * fn)

def compute_ACscore_unidirectional(mapped_labels, direction):
    n = len(mapped_labels)
    ACscore_array = np.full(n + 1, np.nan)

    if direction == "forward":
        predicted = np.ones(n, dtype=int)
    else:
        predicted = np.zeros(n, dtype=int)

    tp, tn, fp, fn = compute_confusion_matrix(mapped_labels, predicted)
    ACscore_array[0] = compute_ac_score(tp, tn, fp, fn)

    for i in range(1, n + 1):
        if direction == "forward":
            tp, tn, fp, fn = update_cm_forward(tp, tn, fp, fn, mapped_labels[i - 1])
        else:
            tp, tn, fp, fn = update_cm_backward(tp, tn, fp, fn, mapped_labels[i - 1])

        ACscore_array[i] = compute_ac_score(tp, tn, fp, fn)

    ACscore = np.nanmax(ACscore_array)
    idx_max = np.nanargmax(ACscore_array)
    return ACscore_array, ACscore, idx_max

def update_cm_forward(tp, tn, fp, fn, label):
    if label == 0:
        tn += 1
        fp -= 1
    else:
        tp -= 1
        fn += 1
    return tp, tn, fp, fn

def update_cm_backward(tp, tn, fp, fn, label):
    if label == 1:
        tp += 1
        fn -= 1
    else:
        tn -= 1
        fp += 1
    return tp, tn, fp, fn

File: matlab/computeACscore.m
function [ACscore, ACscoreArray, bestCut, direction, bestPredicted] = computeACscore(labels, markers, positiveClass)

% ACscore is a balanced metric for the evaluation of model's
% performance of binary classifiers, and for guiding their
% training, which is suitable for unbalanced data as well as on
% balanced data.

% ACscore = 2 * (sensitivity * specificity) / (sensitivity + specificity).

%%% Input  %%%
% labels: Groundtruth labels of data. Be a string cell, char cell, numeric cell,string array, numeric array,or logical array of 1 dimension.
% markers: Either predicted labels or scores from DR results. In the same format as labels.
% positiveClass: Assign one of the two classes as the positive class.

%%% Output %%%
% ACscore: The ACscore evaluation result. If markers are scores, this will be the highest one among the ACscore evaluations for N+1 possible cuts.
% ACscoreArray (only when markers are scores): ACscore evaluations for N+1 possible cuts.
% bestCut (only when markers are scores): The cut corresponding to the higest ACscore from the ACscore array, obtained as the middle point of the pair of scores closest to the cut and lying at each side of the cut (Except the cut is lower than the smallest score or higher than the largest score).
% direction: 1 if scores greater or equal to the bestCut are predicted as positive; 0 if scores greater or equal to the bestCut are predicted as negative;
% bestPredicted (only when markers are scores): The predicted labels corresponding to the higest ACscore evaluation from the ACscore array.

%%% An example to use %%%
% N=1E3;
% rng(10086)
% markers=randperm(N);
% labels=cellstr([repmat({"red"},1,N/2-1),{"green"},{"red"},repmat({"green"},1,N/2-1)]);
% [ACscore, ACscoreArray, bestCut, direction, bestPredicted]=computeACscore(labels, markers,char("green"));
% [~,idx]=sort(markers);
% plot(ACscoreArray(idx));
% xlabel("Marker ranking")
% ylabel("ACscore")
% disp(['ACscore: ', num2str(ACscore)])
% disp(['bestCut: ', num2str(bestCut)])
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


validateattributes(labels,{'cell','numeric','string','logical'},{'vector'})
validateattributes(markers,{'cell','numeric','string','logical'},{'vector'})

if iscell(markers)
    IsMarkerScore = 0;
elseif numel(unique(markers))==2
    IsMarkerScore = 0;
else
    IsMarkerScore = 1;
end

if iscell(labels)
    if isnumeric(labels{1})
        labels = cellstr(string(cell2mat(labels)));
    elseif isstring(labels{1})
        labels = cellstr(labels);
    elseif ~ischar(labels{1})
        error('Input argument must be a string cell, char cell, numeric cell,string array, numeric array,or logical array in a row or column');
    end
    if ~IsMarkerScore
        if isnumeric(markers{1})
            markers = cellstr(string(cell2mat(markers)));
        elseif isstring(markers{1})
            markers = cellstr(markers);
        elseif ~ischar(markers{1})
            error('Input argument must be a string cell, char cell, numeric cell,string array, numeric array,or logical array in a row or column');
        end
    end
end

if size(labels,1)==1
    labels=labels';
    markers=markers';
end

List_labels = unique(labels,'stable');
if numel(List_labels)~=2
    error("The number of classes is not 2")
end
if isempty(intersect(List_labels,positiveClass))
    error("The positive class specified is incorrect")
end

negativeClass=setdiff(List_labels,positiveClass);
S = size(markers);
N = S(S~=1);
mapped_labels = zeros(N,1,'logical');
if ~iscell(labels)
    mapped_labels(labels==positiveClass,1) = 1; 
    mapped_labels(labels==negativeClass,1) = 0; 
    if ~IsMarkerScore
        mapped_markers(markers==positiveClass,1) = 1; 
        mapped_markers(markers==negativeClass,1) = 0;
    end
else 
    if ~iscell(positiveClass)
        positiveClass = {positiveClass};
    end
    mapped_labels(strcmp(labels,positiveClass{:}),1) = 1;
    mapped_labels(strcmp(labels,negativeClass{:}),1) = 0; 
    if ~IsMarkerScore
        mapped_markers(strcmp(markers,positiveClass{:}),1) = 1;
        mapped_markers(strcmp(markers,negativeClass{:}),1) = 0;
    end
end

if ~IsMarkerScore    % for predicted labels
    [tp,tn,fp,fn] = computeCM(mapped_labels,mapped_markers);
    ACscore = docomputeACscore(tp,tn,fp,fn);
else                 % for scores
    [~,idx_sorted] = sort(markers,'ascend');
    [idx_sorted_inv,~,~] = find((1:N)==idx_sorted);


    [ACscoreArrayForward,ACscoreForward,idx_max_forward]=computeACscoreUnidirect(mapped_labels(idx_sorted),'forward');
    [ACscoreArrayBackward,ACscoreBackward,idx_max_Backward]=computeACscoreUnidirect(mapped_labels(idx_sorted),'backward');

    if ACscoreForward >= ACscoreBackward
        ACscoreArray = ACscoreArrayForward(idx_sorted_inv);
        ACscore = ACscoreForward;
        idx_max = idx_max_forward;
        direction = 1;
    else
        ACscoreArray = ACscoreArrayBackward(idx_sorted_inv);
        ACscore = ACscoreBackward;
        idx_max = idx_max_Backward;
        direction = 0;
    end


    Delta = 1E-6;
    if idx_max==1
        bestCut = markers(idx_sorted(1)) - Delta;
    elseif idx_max==N+1
        bestCut = markers(idx_sorted(end)) + Delta;
    else
        bestCut = (markers(idx_sorted(idx_max-1)) + markers(idx_sorted(idx_max))) / 2; 
    end
    if direction == 1
        bestPredicted = [repmat(negativeClass,idx_sorted(idx_max)-1,1);repmat(positiveClass,N-(idx_sorted(idx_max)-1),1)];
    else
        bestPredicted = [repmat(positiveClass,idx_sorted(idx_max)-1,1);repmat(negativeClass,N-(idx_sorted(idx_max)-1),1)];
    end
end
end


function [ACscoreArray,ACscore,idx_max]=computeACscoreUnidirect(mapped_labels,dir)
N=length(mapped_labels);
if strcmp(dir,"forward")
    predicted = ones(N,1,'logical');
else
    predicted = zeros(N,1,'logical');
end
ACscoreArray = nan(N,1);

[tp,tn,fp,fn]=computeCM(mapped_labels,predicted);
ACscoreArray(1) = docomputeACscore(tp,tn,fp,fn);

for i=2:N+1     % N points have N+1 cut candidates
    if strcmp(dir,"forward")
        [tp,tn,fp,fn]=updateCM_forward(tp,tn,fp,fn,mapped_labels,i);
    else
        [tp,tn,fp,fn]=updateCM_backward(tp,tn,fp,fn,mapped_labels,i);
    end
    ACscoreArray(i) = docomputeACscore(tp,tn,fp,fn);
end
[ACscore,idx_max] = max(ACscoreArray);
end


function ACscore = docomputeACscore(tp,tn,fp,fn)
Total = tp+tn+fp+fn;
if (tp+tn)==Total
    ACscore=1;
elseif (fp+fn)==Total
    ACscore=0;
else
    tptn = tp*tn;
    ACscore = 2*tptn / (2*tptn+tp*fp+tn*fn);
end
end


function [tp,tn,fp,fn] = computeCM(labels,predicted)
tp = sum(predicted == 1 & labels == 1);
tn = sum(predicted == 0 & labels == 0);
fp = sum(predicted == 1 & labels == 0);
fn = sum(predicted == 0 & labels == 1);
end

function [tp,tn,fp,fn] = updateCM_forward(tp,tn,fp,fn,mapped_labels,i)
if mapped_labels(i-1)==0
    tn = tn + 1;
    fp = fp -1;
else
    tp = tp - 1;
    fn = fn + 1;
end
end

function [tp,tn,fp,fn] = updateCM_backward(tp,tn,fp,fn,mapped_labels,i)
if mapped_labels(i-1)==1
    tp = tp + 1;
    fn = fn -1;
else
    tn = tn - 1;
    fp = fp + 1;
end
end